README.md
# angular2gridster

[![npm version](https://badge.fury.io/js/angular2gridster.svg)](https://badge.fury.io/js/angular2gridster)

Angular implementation of well known Gridster (no jQuery, no external libraries, only Angular and Rx.js). [Demo](https://swiety85.github.io/angular2gridster/).

1. [Getting started](https://github.com/swiety85/angular2gridster/wiki/Getting-started)
2. [What is Angular2gridster and why to use it?](https://github.com/swiety85/angular2gridster/wiki)
3. [API Documentation](https://github.com/swiety85/angular2gridster/wiki/API-Documentation)
4. [Roadmap](https://github.com/swiety85/angular2gridster/wiki/Roadmap)

More comprehensive documentation is available in [Wiki](https://github.com/swiety85/angular2gridster/wiki).

Development progress can be tracked in [Milestones](https://github.com/swiety85/angular2gridster/milestones) and in [Project board](https://github.com/swiety85/angular2gridster/projects/1).

## Versions:

-   Version **9.x** works with **Angular 9.x**.
-   Version **8.x** works with **Angular 8.x**.
-   Version **7.x** works with **Angular 7.x**.
-   Version **6.x** works with **Angular 6.x**.
-   Version **5.x** works with **Angular 5.x**.
-   Version **4.x** works with **Angular 4.x**.

Versions **1.x** and **0.x** works only with **Angular 4.x**, but the newest states you can find in **v4.x**.

## Installation

```shell
npm install angular2gridster
```

Once installed you need to import our module:

```js
...
import { GridsterModule } from 'angular2gridster';

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    ...
    GridsterModule.forRoot() // .forRoot() is required since version v4+
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

The example it imports in AppModule, but it could also be imported in any other module - depends where you want to use it.

### Additional steps for System.js

To make Angular2gridster works with System.js you need to provide dedicated configuration in `systemjs.config.js`.
It requires change in `map` object and 'packages' as follows:

```js
System.config({
    map: {
        // ...
        rxjs: 'node_modules/rxjs',
        angular2gridster: 'node_modules/angular2gridster',
    },
    packages: {
        // ...
        rxjs: {defaultExtension: 'js'},
        angular2gridster: {main: 'dist/index.js', defaultExtension: 'js'},
    },
});
```

## Example usage

```html
<ngx-gridster [options]="gridsterOptions" [draggableOptions]="{ handlerClass: 'panel-heading' }">
    <ngx-gridster-item
        *ngFor="let widget of widgets"
        [(x)]="widget.x"
        [(y)]="widget.y"
        [(w)]="widget.w"
        [(h)]="widget.h"
    >
        <!-- some content -->
    </ngx-gridster-item>
</ngx-gridster>
```

For version before 6.0.0:

```html
<gridster [options]="gridsterOptions" [draggableOptions]="{ handlerClass: 'panel-heading' }">
    <gridster-item
        *ngFor="let widget of widgets"
        [(x)]="widget.x"
        [(y)]="widget.y"
        [(w)]="widget.w"
        [(h)]="widget.h"
    >
        <!-- some content -->
    </gridster-item>
</gridster>
```

```js
widgets: Array<any> = [...];
gridsterOptions = {
  lanes: 2, // how many lines (grid cells) dashboard has
  direction: 'vertical', // items floating direction: vertical/horizontal/none
  floating: false, // default=true - prevents items to float according to the direction (gravity)
  dragAndDrop: false, // possible to change items position by drag n drop
  resizable: false, // possible to resize items by drag n drop by item edge/corner
  useCSSTransforms: true, // Uses CSS3 translate() instead of position top/left - significant performance boost.
  responsiveSizes: true, // allow to set different item sizes for different breakpoints
  // ResponsiveOptions can overwrite default configuration with any option available for specific breakpoint.
  responsiveOptions: [
        {
            breakpoint: 'sm',
            lanes: 3
        },
        {
            breakpoint: 'md',
            minWidth: 768,
            lanes: 4,
            dragAndDrop: true,
            resizable: true
        },
        {
            breakpoint: 'lg',
            lanes: 6,
            dragAndDrop: true,
            resizable: true
        },
        {
            breakpoint: 'xl',
            minWidth: 1800,
            lanes: 8,
            dragAndDrop: true,
            resizable: true
        }
    ]
};
```

**Warning**

If you use `responsiveOptions`, then item coords will be assigned to different breakpoint attributes:

-   till `sm` (480px), it uses `x` and `y` attributes
-   `sm` (480px - 768px), it uses `xSm` and `ySm` attributes
-   `md` (768px - 1250px), it uses `xMd` and `yMd` attributes
-   `lg` (1250px - 1800px), it uses `xLg` and `yLg` attributes
-   from `xl` (1800px), it uses `xXl` and `yXl` attributes

(widths in px are only example and works for `responsiveOptions in example above).

If you set `responsiveSizes: true`, item size can be different for different breakpoints. In this case size will be binded to following attributes:

-   till `sm` (480px), it uses `w` and `h` attributes
-   `sm` (480px - 768px), it uses `wSm` and `hSm` attributes
-   `md` (768px - 1250px), it uses `wMd` and `hMd` attributes
-   `lg` (1250px - 1800px), it uses `wLg` and `hLg` attributes
-   from `xl` (1800px), it uses `wXl` and `hXl` attributes

## Demo

Clone or download this repository and just run:

```js
npm i
npm run build
npm start
```

Windows:
In CMD run `MKLINK /D .\src\app\angular2gridster D:\your\absolute\path\to\angular2gridster\projects\angular2gridster\src\lib` (make sure the second path is the correct absolute for your environment).

Then run the following:
```shell
    npm install
    ng serve
```

Go to: http://localhost:4200/

## Compilation problems

If somebody will have compilation problems please add an issue (if not yet created). I will try to fix it as soon as possible.
Angular compiler has still some issues opened and it is changing frequently.

As a temporary solution copy files from `/projects/angular2gridster/src/lib` folder to dedicated folder in your project.

## Issues

If the current behavior is a bug or you can illustrate your feature request better with an example,
please provide the steps to reproduce and if possible a minimal demo of the problem via CodeSandbox:

[![Edit Angular](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/angular-otned?fontsize=14)

The project is in development so don't hesitate to writte any questions or suggestion on issue list.
I look forward to get a response from you.

## Origin

This project was created on idea of [GridList](https://github.com/hootsuite/grid). Great alternative for Gridster.

## Compilation to precompiled repository

In order to compile this for use in skyux-lib-reporting-dashboard-editor, follow these steps:

1. `cd` to `out-tsc\lib\projects\angular2gridster\src` and do a `git pull origin master` to ensure the submodule is at latest.
2. `cd` back to root and run `.\node_modules\.bin\tsc --project .\projects\angular2gridster\tsconfig.lib.json`.
3. The generated *.d.ts files in `out-tsc` are missing the decorators (e.g. `@NgModule`, `@NgClass`, `@Directive`, `@Input`) and the parameters for private functions which need to be manually re-added.
4. `cd` to `out-tsc\lib\projects\angular2gridster\src`, commit and push the changes to the angular2gridster-precompiled repo.
5. `cd` back to root and commit and push changes to the angular2gridster repo.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [9.0.0](https://github.com/swiety85/angular2gridster/compare/v8.1.2...v9.0.0) (2020-02-28)


### Features

* Upgrade to Angular 9 ([c3fee22](https://github.com/swiety85/angular2gridster/commit/c3fee22))


### Bug Fixes

* Fix security issues ([6f02e08](https://github.com/swiety85/angular2gridster/commit/6f02e08))
* import style according to angular style guide ([a4004a5](https://github.com/swiety85/angular2gridster/commit/a4004a5))



### [8.1.2](https://github.com/swiety85/angular2gridster/compare/v8.1.1...v8.1.2) (2019-10-23)


### Bug Fixes

* Use rounding again for cell size ([9962aa0](https://github.com/swiety85/angular2gridster/commit/9962aa0)), closes [#347](https://github.com/swiety85/angular2gridster/issues/347)



### [8.1.1](https://github.com/swiety85/angular2gridster/compare/v8.1.0...v8.1.1) (2019-10-10)


### Bug Fixes

* Fix lodash Vulnerable version ([7e25199](https://github.com/swiety85/angular2gridster/commit/7e25199))
* **drag:** Remove rounding in clalculating cell sizes and item positions ([536cd41](https://github.com/swiety85/angular2gridster/commit/536cd41)), closes [#346](https://github.com/swiety85/angular2gridster/issues/346) [#334](https://github.com/swiety85/angular2gridster/issues/334)



## [8.1.0](https://github.com/swiety85/angular2gridster/compare/v8.0.1...v8.1.0) (2019-08-20)


### Bug Fixes

* **demo:** Fix resize widget by buttons in demo ([3aab09c](https://github.com/swiety85/angular2gridster/commit/3aab09c))


### Features

* **config:** add Gridster responsiveToParent option ([35ca28c](https://github.com/swiety85/angular2gridster/commit/35ca28c)), closes [#201](https://github.com/swiety85/angular2gridster/issues/201) [#333](https://github.com/swiety85/angular2gridster/issues/333)



### [8.0.1](https://github.com/swiety85/angular2gridster/compare/v8.0.0...v8.0.1) (2019-06-10)


### Features

* Upgrade to Angular8 ([2bcc03f](https://github.com/swiety85/angular2gridster/commit/2bcc03f))



<a name="7.0.2"></a>
## [7.0.2](https://github.com/swiety85/angular2gridster/compare/v7.0.1...v7.0.2) (2019-01-26)


### Bug Fixes

* **demo:** demo not scrollable content ([1fb0cd4](https://github.com/swiety85/angular2gridster/commit/1fb0cd4))
* **drag:** fix drag mouse position in scrollabe container ([d044b54](https://github.com/swiety85/angular2gridster/commit/d044b54))



<a name="7.0.1"></a>
## [7.0.1](https://github.com/swiety85/angular2gridster/compare/v7.0.0...v7.0.1) (2018-11-06)


### Bug Fixes

* **build:** fix peerDependencies ([d9f6f02](https://github.com/swiety85/angular2gridster/commit/d9f6f02))



<a name="7.0.2"></a>
## [7.0.2](https://github.com/swiety85/angular2gridster/compare/v7.0.1...v7.0.2) (2019-01-26)


### Bug Fixes

* **demo:** demo not scrollable content ([1fb0cd4](https://github.com/swiety85/angular2gridster/commit/1fb0cd4))
* **drag:** fix drag mouse position in scrollabe container ([d044b54](https://github.com/swiety85/angular2gridster/commit/d044b54))



<a name="7.0.1"></a>
## [7.0.1](https://github.com/swiety85/angular2gridster/compare/v7.0.0...v7.0.1) (2018-11-06)


### Bug Fixes

* **build:** fix peerDependencies ([d9f6f02](https://github.com/swiety85/angular2gridster/commit/d9f6f02))



<a name="7.0.0"></a>
# [7.0.0](https://github.com/swiety85/angular2gridster/compare/v6.0.4...v7.0.0) (2018-11-04)


### Features

* Angular 7 support ([38a1be7](https://github.com/swiety85/angular2gridster/commit/38a1be7))



<a name="6.0.4"></a>
## [6.0.4](https://github.com/swiety85/angular2gridster/compare/v6.0.3...v6.0.4) (2018-10-09)


### Bug Fixes

* fix element.style is undefined error ([7baa71c](https://github.com/swiety85/angular2gridster/commit/7baa71c)), closes [#298](https://github.com/swiety85/angular2gridster/issues/298)



<a name="6.0.3"></a>

## [6.0.3](https://github.com/swiety85/angular2gridster/compare/v6.0.2...v6.0.3) (2018-10-08)

### Bug Fixes

-   Expose private exports in bundle ([e50b070](https://github.com/swiety85/angular2gridster/commit/e50b070)), closes [#299](https://github.com/swiety85/angular2gridster/issues/299)
-   Prevent error while dragging widget outside browser window
-   Fix getHProperty function. Fixes #283

<a name="6.0.2"></a>

# [6.0.2](https://github.com/swiety85/angular2gridster/compare/v6.0.1...v6.0.2) (2018-07-29)

### Bug Fixes

-   fix: set default draggableOptions to empty object
-   fix: change event - oldValues empty when value = 0

<a name="6.0.1"></a>

# [6.0.1](https://github.com/swiety85/angular2gridster/compare/v6.0.0-rc.1...v6.0.1) (2018-06-18)

### Bug Fixes

-   fix: position and size binding on outside change
    Optimize amount of render calls

<a name="6.0.0-rc.1"></a>

# [6.0.0-rc.1](https://github.com/swiety85/angular2gridster/compare/v7.0.0-rc.0...v6.0.0-rc.1) (2018-05-29)

### Bug Fixes

-   fix build bundle

<a name="6.0.0-rc.0"></a>

# [6.0.0-rc.0](https://github.com/swiety85/angular2gridster/compare/v5.0.0...v6.0.0-rc.0) (2018-05-29)

### Features

-   **build:** compatibility with Angular 6 ([ceb74e3](https://github.com/swiety85/angular2gridster/commit/ceb74e3))

### BREAKING CHANGES

-   **build:** New bundles works only with Angular 6. New selector names with ngx prefix

<a name="5.0.0"></a>

# [5.0.0](https://github.com/swiety85/angular2gridster/compare/v4.0.0...v5.0.0) (2018-05-26)

### Features

-   angular 5 compatibility ([cec9083](https://github.com/swiety85/angular2gridster/commit/cec9083)), closes [#244](https://github.com/swiety85/angular2gridster/issues/244)

<a name="4.0.0"></a>

# [4.0.0](https://github.com/swiety85/angular2gridster/compare/v1.6.1...v4.0.0) (2018-05-26)

### Features

-   allow drag item prototypes to different gridster components in different modules
    BREAKING CHANGES: requeires .forRoot() in module import

<a name="1.6.0"></a>


out-tsc/lib/src/app/app.component.js
import * as tslib_1 from "tslib";
import { Component, ViewChild, ViewEncapsulation } from '@angular/core';
import { GridsterComponent } from './angular2gridster/gridster.component';
var AppComponent = /** @class */ (function () {
    function AppComponent() {
        this.itemOptions = {
            maxWidth: 3,
            maxHeight: 4
        };
        this.gridsterOptions = {
            // core configuration is default one - for smallest view. It has hidden minWidth: 0.
            lanes: 2,
            direction: 'vertical',
            floating: true,
            dragAndDrop: true,
            resizable: true,
            resizeHandles: {
                s: true,
                e: true,
                se: true
            },
            widthHeightRatio: 1,
            lines: {
                visible: true,
                color: '#afafaf',
                width: 2
            },
            shrink: true,
            useCSSTransforms: true,
            responsiveView: true,
            responsiveDebounce: 500,
            responsiveSizes: true,
            // List of different gridster configurations for different breakpoints.
            // Each breakpoint is defined by name stored in "breakpoint" property. There is fixed set of breakpoints
            // available to use with default minWidth assign to each.
            // - sm: 576 - Small devices
            // - md: 768 - Medium devices
            // - lg: 992 - Large devices
            // - xl: 1200 - Extra large
            // MinWidth for each breakpoint can be overwritten like it's visible below.
            // ResponsiveOptions can overwrite default configuration with any option available.
            responsiveOptions: [
                {
                    breakpoint: 'sm',
                    // minWidth: 480,
                    lanes: 3
                },
                {
                    breakpoint: 'md',
                    minWidth: 768,
                    lanes: 4
                },
                {
                    breakpoint: 'lg',
                    minWidth: 1250,
                    lanes: 6
                },
                {
                    breakpoint: 'xl',
                    minWidth: 1800,
                    lanes: 8
                }
            ]
        };
        this.gridsterDraggableOptions = {
            handlerClass: 'panel-heading'
        };
        this.title = 'Angular2Gridster';
        this.widgetsCopy = [];
        this.widgets = [
            {
                x: 0, y: 0,
                w: 1, h: 2,
                wSm: 1, hSm: 1,
                wMd: 1, hMd: 2,
                wLg: 1, hLg: 1,
                wXl: 2, hXl: 2,
                dragAndDrop: true,
                resizable: true,
                title: 'Basic form inputs 1',
                variableHeight: false,
                contents: []
            },
            {
                x: 1, y: 0,
                w: 1, h: 2,
                wSm: 2, hSm: 1,
                wMd: 2, hMd: 1,
                wLg: 3, hLg: 1,
                wXl: 3, hXl: 1,
                dragAndDrop: true,
                resizable: true,
                title: 'Basic form inputs 2',
                variableHeight: false,
                contents: []
            },
            {
                x: 1, y: 1,
                w: 2, h: 1,
                wSm: 1, hSm: 2,
                wMd: 1, hMd: 2,
                wLg: 2, hLg: 1,
                wXl: 3, hXl: 1,
                dragAndDrop: true,
                resizable: true,
                title: 'Basic form inputs 3',
                variableHeight: true,
                contents: []
            },
            {
                x: 3, y: 1,
                w: 1, h: 2,
                wSm: 1, hSm: 2,
                wMd: 1, hMd: 2,
                wLg: 3, hLg: 1,
                wXl: 3, hXl: 1,
                dragAndDrop: true,
                resizable: true,
                title: 'Basic form inputs 4',
                contents: []
            },
            {
                w: 1, h: 2,
                wSm: 1, hSm: 2,
                wMd: 1, hMd: 2,
                wLg: 3, hLg: 1,
                wXl: 3, hXl: 1,
                dragAndDrop: true,
                resizable: true,
                title: 'Basic form inputs x',
                contents: []
            }
        ];
    }
    AppComponent_1 = AppComponent;
    AppComponent.prototype.ngOnInit = function () {
        this.widgetsCopy = this.widgets.map(function (widget) { return (tslib_1.__assign({}, widget)); });
    };
    AppComponent.prototype.onReflow = function (event) {
        console.log('onReflow', event);
    };
    AppComponent.prototype.removeLine = function (gridster) {
        gridster.setOption('lanes', --this.gridsterOptions.lanes)
            .reload();
    };
    AppComponent.prototype.getTitle = function () {
        return this.title;
    };
    AppComponent.prototype.addLine = function (gridster) {
        gridster.setOption('lanes', ++this.gridsterOptions.lanes)
            .reload();
    };
    AppComponent.prototype.setWidth = function (widget, size, e, gridster) {
        e.stopPropagation();
        e.preventDefault();
        if (size < 1) {
            size = 1;
        }
        widget.w = size;
        gridster.reload();
        return false;
    };
    AppComponent.prototype.setHeight = function (widget, size, e, gridster) {
        e.stopPropagation();
        e.preventDefault();
        if (size < 1) {
            size = 1;
        }
        widget.h = size;
        gridster.reload();
        return false;
    };
    AppComponent.prototype.optionsChange = function (options) {
        this.gridsterOptions = options;
        console.log('options change:', options);
    };
    AppComponent.prototype.swap = function () {
        this.widgets[0].x = 3;
        this.widgets[3].x = 0;
    };
    AppComponent.prototype.addWidgetFromDrag = function (gridster, event) {
        var item = event.item;
        var breakpoint = gridster.options.breakpoint;
        var widget = {
            dragAndDrop: true,
            resizable: true,
            title: 'New widget'
        };
        widget[AppComponent_1.W_PROPERTY_MAP[breakpoint] || 'w'] = item.w;
        widget[AppComponent_1.H_PROPERTY_MAP[breakpoint] || 'h'] = item.h;
        widget[AppComponent_1.X_PROPERTY_MAP[breakpoint] || 'x'] = item.x;
        widget[AppComponent_1.Y_PROPERTY_MAP[breakpoint] || 'y'] = item.y;
        for (var _i = 0, _a = ['wSm', 'hSm', 'wMd', 'hMd', 'wLg', 'hLg', 'wXl', 'hXl']; _i < _a.length; _i++) {
            var rwdProp = _a[_i];
            if (event.item.itemPrototype.hasOwnProperty(rwdProp)) {
                widget[rwdProp] = event.item.itemPrototype[rwdProp];
            }
        }
        this.widgets.push(widget);
        console.log('add widget from drag to:', gridster);
    };
    AppComponent.prototype.over = function (event) {
        var size = event.item.calculateSize(event.gridster);
        event.item.itemPrototype.$element.querySelector('.gridster-item-inner').style.width = size.width + 'px';
        event.item.itemPrototype.$element.querySelector('.gridster-item-inner').style.height = size.height + 'px';
        event.item.itemPrototype.$element.classList.add('is-over');
    };
    AppComponent.prototype.out = function (event) {
        event.item.itemPrototype.$element.querySelector('.gridster-item-inner').style.width = '';
        event.item.itemPrototype.$element.querySelector('.gridster-item-inner').style.height = '';
        event.item.itemPrototype.$element.classList.remove('is-over');
    };
    AppComponent.prototype.addWidgetWithoutData = function () {
        this.widgets.push({
            title: 'Basic form inputs X',
            dragAndDrop: true,
            resizable: true,
            content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et ' +
                'dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea ' +
                'commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla ' +
                'pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est ' +
                'laborum.'
        });
    };
    AppComponent.prototype.addWidget = function (gridster) {
        this.widgets.push({
            x: 4, y: 0, w: 1, h: 1,
            dragAndDrop: true,
            resizable: true,
            title: 'Basic form inputs 5',
            content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et ' +
                'dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea ' +
                'commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla ' +
                'pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est ' +
                'laborum.'
        });
        console.log('widget push', this.widgets[this.widgets.length - 1]);
    };
    AppComponent.prototype.remove = function ($event, index, gridster) {
        $event.preventDefault();
        this.widgets.splice(index, 1);
        console.log('widget remove', index);
    };
    AppComponent.prototype.removeAllWidgets = function () {
        this.widgets = [];
    };
    AppComponent.prototype.itemChange = function ($event, gridster) {
        console.log('item change', $event);
    };
    AppComponent.prototype.resetWidgets = function () {
        this.widgets = this.widgetsCopy.map(function (widget) { return (tslib_1.__assign({}, widget)); });
    };
    AppComponent.X_PROPERTY_MAP = {
        sm: 'xSm',
        md: 'xMd',
        lg: 'xLg',
        xl: 'xXl'
    };
    AppComponent.Y_PROPERTY_MAP = {
        sm: 'ySm',
        md: 'yMd',
        lg: 'yLg',
        xl: 'yXl'
    };
    AppComponent.W_PROPERTY_MAP = {
        sm: 'wSm',
        md: 'wMd',
        lg: 'wLg',
        xl: 'wXl'
    };
    AppComponent.H_PROPERTY_MAP = {
        sm: 'hSm',
        md: 'hMd',
        lg: 'hLg',
        xl: 'hXl'
    };
    tslib_1.__decorate([
        ViewChild(GridsterComponent),
        tslib_1.__metadata("design:type", GridsterComponent)
    ], AppComponent.prototype, "gridster", void 0);
    AppComponent = AppComponent_1 = tslib_1.__decorate([
        Component({
            selector: 'demo-root',
            templateUrl: './app.component.html',
            styleUrls: ['./app.component.css'],
            encapsulation: ViewEncapsulation.None
        })
    ], AppComponent);
    return AppComponent;
    var AppComponent_1;
}());
export { AppComponent };
//# sourceMappingURL=app.component.js.map